### PR TITLE
[mlir][ArmSME] Add rewrites to swap extract of extend

### PR DESCRIPTION
**Pull request by c-rhodes**
In mixed matmul lowering (e.g., i8 to i32) we're seeing the following sequence:

  %0 = arith.extsi %src : vector<4x[8]xi8> to vector<4x[8]xi32>
  %1 = vector.extract %0[0] : vector<[8]xi32> from vector<4x[8]xi32>
  %lhs = vector.scalable.extract %1[0] : vector<[4]xi32> from vector<[8]xi32>

  ... (same for rhs)

  %2 = vector.outerproduct %lhs, %rhs, %acc vector<[4]xi32>, vector<[4]xi32>

  // x4 chained by accumulator

This chain of 4 outer products can be fused into a single 4-way widening variant but the pass doesn't match on the IR, as it expects the source of the inputs to be an extend and it can't look through the extracts.

This patch fixes this with two rewrites that swaps extract(extend) into extend(extract).

Related to #78975, #79288.

### CODE REVIEW

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 1]
```diff
@@ -338,6 +338,105 @@ struct LegalizeTransferWriteOpsByDecomposition
   }
 };
 
+// Shuffles arith extend ops after vector.extract op.
```
Comment (banach-space) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:341]: [nit] I find this form particularly easy to parse 😅 
```suggestion
//  arith.extend(vector.extract) -> vector.extract(arith.extend) 
```

File: mlir/test/Dialect/ArmSME/vector-legalization.mlir
Comment (banach-space) [inline mlir/test/Dialect/ArmSME/vector-legalization.mlir:1]: Perhaps add some negative tests and a test for different extend type? (unsigned? fp?) Definitely no need to test the whole "matrix" though :)

**Review by banach-space — APPROVED** [review]
Makes sense to me, thanks! Just a few small suggestions.

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 2]
```diff
@@ -338,6 +338,105 @@ struct LegalizeTransferWriteOpsByDecomposition
...
+
+    // Create new extract from source of extend.
+    Value newExtract = rewriter.create<vector::ExtractOp>(
+        loc, extendSource, extractOp.getMixedPosition());
+
+    // Extend new extract to original result type.
+    Operation *newExtend =
+        rewriter.create(loc, extendOpName, Value(newExtract), resultType);
+
+    rewriter.replaceOp(extractOp, newExtend->getResult(0));
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:386]:
```suggestion
    rewriter.replaceOp(extractOp, newExtend);
```

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 3]
```diff
@@ -338,6 +338,105 @@ struct LegalizeTransferWriteOpsByDecomposition
...
+
+    Value extendSource = extendOp->getOperand(0);
+    StringAttr extendOpName = extendOp->getName().getIdentifier();
+    VectorType extendSourceVectorType =
+        cast<VectorType>(extendSource.getType());
+
+    // Create new extract from source of extend.
+    VectorType extractResultVectorType =
+        VectorType::Builder(resultType)
+            .setElementType(extendSourceVectorType.getElementType());
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:426]:
```suggestion
    VectorType extractResultVectorType = resultType.clone(extendSourceVectorType.getElementType());
```

File: mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp [hunk 4]
```diff
@@ -338,6 +338,105 @@ struct LegalizeTransferWriteOpsByDecomposition
...
+    Value newExtract = rewriter.create<vector::ScalableExtractOp>(
+        loc, extractResultVectorType, extendSource, extractOp.getPos());
+
+    // Extend new extract to original result type.
+    Operation *newExtend =
+        rewriter.create(loc, extendOpName, Value(newExtract), resultType);
+
+    rewriter.replaceOp(extractOp, newExtend->getResult(0));
```
Comment (MacDue) [inline mlir/lib/Dialect/ArmSME/Transforms/VectorLegalization.cpp:434]:
```suggestion
    rewriter.replaceOp(extractOp, newExtend);
```

**Review by MacDue — APPROVED** [review]
The rewrites look good, though I wonder if it'd make more sense to include these in the `-arm-sme-outer-product-fusion` pass instead? They'd work just as well there, and that would also ensure no pass between `-arm-sme-vector-legalization` and `-arm-sme-outer-product-fusion` moves them again (due to some random canonicalization).

**Comment by c-rhodes** [comment]
> The rewrites look good, though I wonder if it'd make more sense to include these in the `-arm-sme-outer-product-fusion` pass instead? They'd work just as well there, and that would also ensure no pass between `-arm-sme-vector-legalization` and `-arm-sme-outer-product-fusion` moves them again (due to some random canonicalization).

Good point, I agree, will move them.

**Comment by c-rhodes** [comment]
> > The rewrites look good, though I wonder if it'd make more sense to include these in the `-arm-sme-outer-product-fusion` pass instead? They'd work just as well there, and that would also ensure no pass between `-arm-sme-vector-legalization` and `-arm-sme-outer-product-fusion` moves them again (due to some random canonicalization).
> 
> Good point, I agree, will move them.

Done 👍

**Pull request merged**